Sara/tec 319 rsa 2025 click to fix with breaking change guidance (#2112)
semgrep/semgrep-docs: docs/semgrep-supply-chain/upgrade-guidance.md

## Review a finding's upgrade guidance 

<ViewDetailsSsc />

![SSC details that provide upgrade guidance.](/img/vuln-panels-ssc.png)
_**Figure**. Useful details that provide upgrade guidance._
<dl>
<dt>A - Upgrade badge</dt>
<dd>Indicates if an upgrade is safe or may break your codebase.</dd>
<dt>B - The line of code (LOC) of the finding</dt>
<dd>This shows the LOC that caused the finding; this does <strong>not</strong> show the LOC where the breaking changes occur.</dd>
<dt>C - Link to change list drawer</dt>
<dd>Click this link to display the LOC where a breaking change may occur.</dd>
<dt>D - Open fix PR button</dt>
<dd>Click this button to open a PR with the code to upgrade the dependency to a safe version, if any.</dd>
</dl>

![Drawer showing all lines of code that must be changed](/img/upgrade-guidance-drawer.png#md-width)
_**Figure**. Drawer showing all the lines of code that must be changed or are safe._

## Create a pull request with fixes

1. Navigate to the **Details** page of the finding for which you want to make a pull request.
1. Click **Fix** > **Open fix PR**.

A pull request includes:

> - The code change to upgrade the dependency

Reviewer: I might phrase this as something more like "the manifest/lockfile change to upgrade the dependency", since we aren't making actual _code_ changes

---
slug: upgrade-guidance
title: Upgrade guidance
hide_title: true
description: Know if a vulnerable package or dependency can be easily and reliably upgraded to a fixed version.
tags:
  - Semgrep Supply Chain
  - Semgrep AppSec Platform
---

import ViewDetailsSsc from "/src/components/procedure/_view-details-ssc.md"
 
# Upgrade guidance

Know if you or your developers can safely and reliably update a vulnerable package or dependency to a fixed version. From there, choose to:

- Have Semgrep open a pull request (PR)
- Create a Jira ticket
- Set the finding's triage status as **To fix**

> Semgrep's dependency upgrade guidance uses AI to analyze if a finding can be **safely upgraded** or if upgrading the package can cause **breaking changes**. Semgrep can also create a PR to upgrade the package.

Reviewer 2: ```suggestion
Semgrep's dependency upgrade guidance uses [Assistant](/semgrep-assistant/overview) to analyze if a finding can be **safely upgraded** or if upgrading the package can cause **breaking changes**. Semgrep can also create a PR to upgrade the package.
```

Feel free to disregard if you think appropriate, but I thought it would be a good time to cross-link to Assistant

## Feature maturity

This feature is in **private beta**. To join the beta, reach out to the [Semgrep support team](/support).

## Supported languages and package managers

- **Python** codebases with the following package managers:
  - `pip`
  - `pip-tools`
  - `pipenv`
  - Poetry

> Semgrep supports the following source code managers:

Reviewer 2: I suggest omitting this sentence since we don't have one for Python and I think the paragraph break might be sufficient?

- GitHub Cloud and GitHub Enterprise Server (self-hosted)
  - This **includes** projects added to Semgrep through Semgrep Managed Scans 

## Prerequisites

> - At least one repository that scans for dependencies through Semgrep Supply Chain. See [Scan third-party dependencies](/semgrep-supply-chain/getting-started).

Reviewer 2: ```suggestion
- At least one repository that [scans for dependencies through Semgrep Supply Chain](/semgrep-supply-chain/getting-started).
```

- Semgrep Assistant must be [enabled](/semgrep-assistant/getting-started).

> - Your **private** GitHub Semgrep app must have **Read and write** access on the **Contents** permission. Refer to the following section for steps.

Reviewer 2: Maybe add an anchor link and remove the last sentence just in case we ever reorganize the page?

> ### Grant **Read and write** access

Reviewer 2: ```suggestion
### Grant **Read and write** access to your content
```

Author: Maybe to "to the private Semgrep GitHub app"

Reviewer 2: Yeah, that works! I just wanted it to be clear what it's for

To change your Semgrep app's permission:

> 1. Navigate to the settings page of your private Semgrep GitHub app; refer to [<i class="fas fa-external-link fa-xs"></i> Changing the permissions of a GitHub app](https://docs.github.com/en/apps/maintaining-github-apps/modifying-a-github-app-registration#changing-the-permissions-of-a-github-app) for steps.

Reviewer 2: ```suggestion
1. Navigate to the settings page of your private Semgrep GitHub app; refer to [<i class="fas fa-external-link fa-xs"></i> Changing the permissions of a GitHub app](https://docs.github.com/en/apps/maintaining-github-apps/modifying-a-github-app-registration#changing-the-permissions-of-a-github-app) for instructions.
```

1. In the **Repository permissions** section, search for `Contents`.
1. Click the drop-down menu and select **Read and write**.

## How it works

After enabling dependency guidance, Semgrep performs post-scan analysis and marks applicable findings as **Safe to upgrade** or with **Breaking changes**.

> - Frequency: this analysis is performed every **two hours** on the latest **full scan**.

Reviewer 2: ```suggestion
- This analysis is performed every **two hours** on the latest **full scan**.
```

> - Only findings **with fixed versions** are marked by Semgrep as **Safe to upgrade** or with **Breaking changes**.

Reviewer 2: What does fixed versions mean? Pinned versions?

Author: Pinned version: they're keeping it at that version and not upgrading it
Fixed version: version where the vulnerability has been fixed / patched

Will edit this to make it clear.

- Findings without any fixed versions have no badge; instead they say **no patch available**.
  ![Finding with no fixed version available](/img/no-patch-available.png#md-width)
  _**Figure**. **Details** page for a finding that has no available fix._

The following chart shows the steps Semgrep performs from scanning to analysis and the actions you can take based on the advice it shows.

![Flowchart explaining how Semgrep provides upgrade guidance and possible actions to take based on its advice.](/img/upgrade-guidance-flowchart.png)

## Review a finding's upgrade guidance 

<ViewDetailsSsc />

![SSC details that provide upgrade guidance.](/img/vuln-panels-ssc.png)
_**Figure**. Useful details that provide upgrade guidance._
<dl>
<dt>A - Upgrade badge</dt>
<dd>Indicates if an upgrade is safe or may break your codebase.</dd>
<dt>B - The line of code (LOC) of the finding</dt>
<dd>This shows the LOC that caused the finding; this does <strong>not</strong> show the LOC where the breaking changes occur.</dd>
<dt>C - Link to change list drawer</dt>

> <dd>Clicking this link displays the LOC where a breaking change may occur.</dd>

Reviewer 2: ```suggestion
<dd>Click this link to display the LOC where a breaking change may occur.</dd>
```
